hax/hax/motr/__init__.py

```diff
@@ -148,52 +162,94 @@ def _entrypoint_request_cb(self, reply_context: Any, req_id: Any,
+        # the same. Thus, invoke send_entrypoint_reply directly.
+        self.send_entrypoint_request_reply(entrypoint_req)
 
     def send_entrypoint_request_reply(self, message: EntrypointRequest):
         reply_context = message.reply_context
         req_id = message.req_id
         remote_rpc_endpoint = message.remote_rpc_endpoint
         process_fid = message.process_fid
+        e_rc = EAGAIN
 
         LOG.debug('Processing entrypoint request from remote endpoint'
                   " '{}', process fid {}".format(remote_rpc_endpoint,
                                                  str(process_fid)))
         sess = principal_rm = confds = None
         try:
             util = self.consul_util
-            sess = util.get_leader_session_no_wait()
-            principal_rm = util.get_session_node(sess)
-            confds = util.get_confd_list()
-            rm_fid = util.get_rm_fid()
+            # When stopping, there's a possibility that hax may receive
+            # an entrypoint request from motr land. In order to unblock
+            # motr land, reply with entrypoint request with no confds
+            # and RM endpoints as the processes might have already
+            # stopped.
+            rc_quorum = 0
+            rm_fid = Fid(0, 0)
+            if self.is_stopping:
+                confds = []
+            else:
+                sess = util.get_leader_session_no_wait()
+                principal_rm = util.get_session_node(sess)
+                confds = util.get_confd_list()
+
+            # Hax may receive entrypoint requests multiple times during its
+            # lifetime. Hax starts motr rconfc to invoke spiel commands. Motr
+            # rconfc establishes connection with principal RM, in case of
+            # principal RM failure, rconfc invalidates its confc and again
+            # requests entrypoint in a hope that there will be another confd
+            # and principal RM elected so that rconfc can resume its
+            # functionality. During shutdown, when each motr process stops,
+            # including confds, hax broadcasts M0_NC_FAILED event for every
+            # STOPPED or FAILED motr process. Motr rconfc on receiving the
+            # failed events for confds, goes re-requests entrypoint information
+            # and this goes on in a loop. In order to break this loop, the
+            # the entrypoint reply must only report alive confds and rm
+            # endpoints. While doing this we need to handle the bootstrapping
+            # case, so we wait until bootstrapping is done that is all the
+            # motr services are up, we check the confd status and exclude
+            # corresponding confd from the entrypoint reply.
+            active_confds = []
+            if self.spiel_ready:
+                for confd in confds:
+                    if not util.is_confd_failed(confd.fid):
+                        active_confds.append(confd)
+                confds = active_confds
```

Comment:
@mssawant : I think this may cause issues when a confd (out of 3) is stopped/failed and then another motr entity (may be due to restart) requests entrypoint from Hax, in this case hax will send two confd's and a ne principal RM. But here **quorum is not possible**, thus failed confd too needs to be passed and make sure that failed  confd is not set as principal RM.

Reply:
Good point!

@mssawant, I think the logic of removing confds from entry-point replies has to be ON only during the shutdown.

Reply:
@madhavemuri, yes, agree, but problem of rconfc happens even before hax shutdown is initiated. So when hax starts sending M0_NC_FAILED requests to itself and whichever alive motr processes, entrypointclient in hax's motr land starts trying to reconnect and starts sending entrypoint requests in a loop. Hax is not stopping yet, so we don't have a flag or condition that can be set here. Thus, we are checking motr statuses before replying to entrypoint requests.

Reply:
Now ideally, if all the processes are marked as failed rconfc must not send any entrypoint requests. But thats not how rconfc works right, it tries in a hope that at some point atleast one of the confds can come back up and rconfc will get the entrypoint reply. If you can let me know how to break this loop then I can happily revert this change.

Reply:
> Hax is not stopping yet, so we don't have a flag or condition that can be set here.

But hax do know that it's going to stop, right? (Because this all happens when we are trying to stop it.) So why doesn't it check and remove confds from entrypoint replies only when it's going to stop?

Reply:
for e.g., I tried not filtering out failed confds at all, in that case rconfc_stop hangs and hax times out.
rconfc after receiving failed event for RC, sends entrypoint request to hax, now if the remaining confds have not stopped yet and entrypoint is replied and hax sends failed event after that, it does not help as rconfc already has started re-initialization and has sent RM borrow request, which hangs rconfc stop later.
```
May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:37,954 [DEBUG] {qconsumer} Broadcasting HA states [HAState(fid=0x7200000000000001:0x9, status=STOPPED), HAState(fid=0x7200000000000001:0x37, status=OFFLINE), HAState(fid=0x7200000000000001:0x65, status=STOPPED)] over ha_link
May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:37,959 [DEBUG] {qconsumer} Process fid=0x7200000000000001:0x9 encloses 2 services as follows: [FidWithType(fid=0x7300000000000001:0xb, service_type='confd'), FidWithType(fid=0x7300000000000001:0xa, service_type='rms')]
May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:37,965 [DEBUG] {qconsumer} Process fid=0x7200000000000001:0x65 encloses 2 services as follows: [FidWithType(fid=0x7300000000000001:0x67, service_type='confd'), FidWithType(fid=0x7300000000000001:0x66, service_type='rms')]
May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: motr[11622]:  b900   WARN  [ha/halon/interface.c:1111:m0_halon_interface_send]  hl=0x7fd904009360 ep=10.230.250.53@tcp:12345:1:1 epoch=0 tag=71 type=2
May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: motr[11622]:  b900   WARN  [ha/halon/interface.c:1111:m0_halon_interface_se_send]  hl=0x7fd904046150 ep=10.230.250.53@tcp:12345:2:1 epoch=0 tag=65 type=2
May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: motr[11622]:  27e0   WARN  [ha/halon/interface.c:335:halon_interface_process_failure_check]  disconnecting all links to no_id=<7200000000000001:9>
May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: motr[11622]:  27e0   WARN  [ha/halon/interface.c:335:halon_interface_process_failure_check]  disconnecting all links to no_id=<7200000000000001:65>

May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:37,977 [DEBUG] {Dummy-1} Received entrypoint request from remote endpoint '10.230.250.53@tcp:12345:1:1'
May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:37,977 [DEBUG] {Dummy-1} enqueue entrypoint request for 10.230.250.53@tcp:12345:1:1
May 26 12:55:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:37,977 [DEBUG] {Dummy-1} Processing entrypoint request from remote endpoint '10.230.250.53@tcp:12345:1:1'
May 26 12:55:38 ssc-vm-1624.colo.seagate.com hare-hax[11622]: motr[11622]:  1900   WARN  [ha/halon/interface.c:1073:m0_halon_interface_entrypoint_reply]  req_id=0:d rc=0 confd_nr=3 confd_quorum=2 rm_fid=<7300000000000001:a> rm_eps=10.230.250.51@tcp:12345:2:1

May 26 12:55:39 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:39,171 [DEBUG] {qconsumer} Setting process status in KV: ssc-vm-1624.colo.seagate.com/processes/0x7200000000000001:0x37:{"state": "M0_CONF_HA_PROCESS_STOPPED"}
May 26 12:55:39 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:39,176 [DEBUG] {qconsumer} Broadcasting HA states [HAState(fid=0x7200000000000001:0x37, status=FAILED)] over ha_link
May 26 12:55:39 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:39,182 [DEBUG] {qconsumer} Process fid=0x7200000000000001:0x37 encloses 2 services as follows: [FidWithType(fid=0x7300000000000001:0x39, service_type='confd'), FidWithType(fid=0x7300000000000001:0x38, service_type='rms')]
May 26 12:55:39 ssc-vm-1624.colo.seagate.com hare-hax[11622]: motr[11622]:  b300   WARN  [ha/halon/interface.c:1111:m0_halon_interface_send]  hl=0x7fd904009360 ep=10.230.250.53@tcp:12345:1:1
May 26 12:55:39 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 12:55:39,182 [DEBUG] {qconsumer} Broadcast HA state complete with the following message_ids = [MessageId(0x7fd904009360, 75)]

May 26 13:00:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 13:00:37,273 [INFO] {Dummy-1} Delivered to endpoint'10.230.250.
May 26 13:00:37 ssc-vm-1624.colo.seagate.com hare-hax[11622]: 2021-05-26 13:00:37,273 [DEBUG] {qconsumer} Stopping rconfc
May 26 13:02:07 ssc-vm-1624.colo.seagate.com systemd[1]: hare-hax.service stop-sigterm timed out. Killing.
May 26 13:02:07 ssc-vm-1624.colo.seagate.com systemd[1]: hare-hax.service: main process exited, code=killed, status=9/KILL
May 26 13:02:07 ssc-vm-1624.colo.seagate.com systemd[1]: Stopped HAX daemon for Hare.
May 26 13:02:07 ssc-vm-1624.colo.seagate.com systemd[1]: Unit hare-hax.service entered failed state.
May 26 13:02:07 ssc-vm-1624.colo.seagate.com systemd[1]: hare-hax.service failed.

```

m0trace
```
163461  31519887.955.007864  77fc0e0  CALL    rm/rm.c:2650:borrow_send                            > incoming: 0x7fd8e8010460 credit: 1
407317  31520650.561.903180  effb960  CALL    spiel/spiel.c:93:m0_spiel_rconfc_stop               >
407329  31520650.561.904414  d3b5930  CALL    fop/fop.c:182:m0_fop_put                            > fop: 0x7fd8f00601d0 HA link msg fop

422945  31520848.466.817366  d3b5930  CALL    fop/fop.c:182:m0_fop_put                            > fop: 0x7fd99805cb50 Credit Borrow [5422946  31520848.466.818890  d3b5930  CALL    fop/fop.c:188:m0_fop_put                            <
422947  31520848.466.823502  d3b5950  DEBUG   rpc/session.c:797:m0_rpc_session_release            session 0x7fd8e800fed8 1 -> 0
422948  31520848.466.827382  d3b58f0  INFO    rpc/session.c:137:session_state_set                 Session 0x7fd8e800fed8: Busy -> Idle
422949  31520848.466.829924  d3b5910  DEBUG   rpc/item.c:737:m0_rpc_item_change_state             0x7fd99805cb70 [REQUEST/M0_RM_FOP_BORROW(88)] xid=1 SENT -> WAITING_FOR_REPLY sm=Outgoing-RPC-Item-sm
```

Reply:
I thought this was due to Hax is not sending FAILED event as part of nvec reply.  Have you fixed note vec part and checked it ?

Reply:
Yes, that is fixed. Now hax is fetching processes and services statuses from Consul and reporting in nvec reply.
The problem here is when nvec is replied, it is possible that not all the processes have stopped. Thus the nvec reply can still contain M0_NC_ONLINE for some of the services.

Reply:
Mandar, it's not the only problem here. Going back to the beginning of this thread... Currently, the semantics of entrypoint reply implies that all configured confds are listed there, not only those which are online. If we are about to break/change this semantics for the normal cluster work (i.e. not shutdown) - we should carefully revise the rconfc logic and test all the possible scenarios.

For example, consider that one of 3 confds goes down (the main one) and rconfc in all processes now gets reply with only 2 confds which are online. Now, when the failed confd restarts - how will it be updated in rconfcs list in all processes?

Reply:
@andriytk, whenever principal rm confd fails, rconfc invalidates and restarts. As part of restarting it requests new entrypoint information, in response to this new entrypoint request, rconfc will get the restarted confd.
But in general, we need not filter out confds and everything must work fine after marking the respective confds and their services as FAILED. But this is not the case presently. 

This is what is happening,
Tried again without filtering out the failed/stopped confds,
M0_NC_FAILED, was sent in nvec reply to confd,
```
324770  31728444.966.378542  0ff86d0  DEBUG   ha/note.c:225:m0_ha_msg_accept                      nv_note[0]=(no_id=<7200000000000001:9>
g_accept                      nv_note[1]=(no_id=<7300000000000001:a> no_state=2)
g_accept                      nv_note[2]=(no_id=<7200000000000001:65> no_state=2)
g_accept                      nv_note[3]=(no_id=<7200000000000001:37> no_state=2)
g_accept                      nv_note[4]=(no_id=<7400000000000001:0> no_state=1)
_accept                       > confc=0x16dd548 note->nv_nr=5
_accept                       nv_note[0]=(no_id=<7200000000000001:9> no_state=2) obj=0x7f633404b6a0 obj->co_status=2
nfc_herd_link__on_death_cb    > lnk=0x7f6334045730 ep:10.230.250.51@tcp:12345:2:1 rc 0 state:0
post                          grp=0x173b3e0 forkq_push: 0x7f6334046e38 -> 0x7f634184b480
nfc_herd_link__on_death_cb    < lnk=0x7f6334045730 co_ha_state = M0_NC_FAILED, return false
_accept                       nv_note[1]=(no_id=<7300000000000001:a> no_state=2) obj=0x7f6334043b40 obj->co_status=2
_accept                       nv_note[2]=(no_id=<7200000000000001:65> no_state=2) obj=0x7f633405b0e0 obj->co_status=2
nfc_herd_link__on_death_cb    > lnk=0x7f6334053520 ep:10.230.250.48@tcp:12345:2:1 rc 0 state:0
post                          grp=0x173b3e0 forkq_push: 0x7f6334054c28 -> 0x7f6334046e38
nfc_herd_link__on_death_cb    < lnk=0x7f6334053520 co_ha_state = M0_NC_FAILED, return false
_accept                       nv_note[3]=(no_id=<7200000000000001:37> no_state=2) obj=0x7f63340533f0 obj->co_status=2
nfc_herd_link__on_death_cb    > lnk=0x7f633404b9a0 ep:10.230.250.53@tcp:12345:2:1 rc 0 state:2
post                          grp=0x173b3e0 forkq_push: 0x7f633404d0a8 -> 0x7f6334054c28
nfc_herd_link__on_death_cb    < lnk=0x7f633404b9a0 co_ha_state = M0_NC_FAILED, return false
_accept                       nv_note[4]=(no_id=<7400000000000001:0> no_state=1) obj=0x7f6288003790 obj->co_status=0
```
still borrow fop was sent,
```
327468  31728444.973.059312  07f7500  CALL    rpc/item.c:1141:m0_rpc_item_send                    > 0x7f633405bb20 [REQUEST/M0_RM_FOP_BORROW(88)] dest_ep=10.230.250.51@tcp:12345:2:1 ri_session=0x7f6288018
```
This timed out rconfc stop
Here is where the borrow request is sent,
```
327341  31728444.972.843844  07f7c40  CALL    rm/rm.c:915:m0_rm_owner_creditor_reset              <
327342  31728444.972.846292  0ff8700  CALL    ha/halon/interface.c:535:halon_interface_link_disconnected_cb   > hii=0x16ceac0 ha=0x16d82
327343  31728444.972.846372  07f7c90  CALL    conf/rconfc.c:1053:rlock_ctx_creditor_setup         < rc=0
327344  31728444.972.846734  07f7c90  CALL    conf/rconfc.c:1873:rconfc_read_lock_get             > rconfc = 0x16dc530
327345  31728444.972.847666  0ff8700  WARN    ha/halon/interface.c:537:halon_interface_link_disconnected_cb   hl=0x7f629c048000 ep=10.23
327346  31728444.972.866520  07f7c40  DEBUG   conf/rconfc.c:1189:rconfc_state_set                 rconfc: 0x16dc530, M0_RCS_CREDITOR_SET -> M0_RCS_GET_RLOCK
327347  31728444.972.868874  07f7c40  CALL    rm/rm_rwlock.c:554:m0_rm_rwlock_req_init            >
327348  31728444.972.869342  07f7bd0  CALL    rm/rm.c:1066:m0_rm_incoming_init                    > incoming: 0x7f6288019090 for owner: 
327349  31728444.972.871446  07f7b80  CALL    rm/rm.c:967:m0_rm_credit_init                       > credit: 0x7f6288019110 with owner: 
0327350  31728444.972.875430  07f7b80  CALL    rm/rm.c:981:m0_rm_credit_init                       <
327351  31728444.972.876738  07f7bd0  CALL    rm/rm.c:1078:m0_rm_incoming_init                    <
327352  31728444.972.877010  07f7c40  CALL    rm/rm_rwlock.c:563:m0_rm_rwlock_req_init            <
327353  31728444.972.878918  07f7c30  CALL    rm/rm.c:1763:m0_rm_credit_get                       > owner: 0x7f6288018cc0 credit: 1
327354  31728444.972.879498  07f7ba0  DEBUG   sm/sm.c:158:m0_sm_asts_run                          grp=0x7f62880194d0 forkq_pop: 0x7f6341
327355  31728444.972.880886  07f7890  CALL    rm/rm.c:1947:owner_balance
327357  31728444.972.883752  07f7890  CALL    rm/rm.c:2048:incoming_check                         >
327358  31728444.972.883888  07f7840  CALL    rm/rm.c:967:m0_rm_credit_init                       > credit: 0x7f63307f7b90 with owner: 0327359  31728444.972.884044  07f7840  CALL    rm/rm.c:981:m0_rm_credit_init                       <
327360  31728444.972.884808  07f7890  CALL    rm/rm.c:2359:incoming_check_with                    > incoming: 0x7f6288019090 credit: 1
327361  31728444.972.885374  07f7890  CALL    rm/rm.c:2650:borrow_send                            > incoming: 0x7f6288019090 credit: 1
```
This blocks rconfc stop.
I will create another ticket to investigate why we need to filter out confds and why it just does not work with notifying M0_NC_FAILED for the services.
One descrepency that I see here is motr is not requesting state of confd service in the nvec request but just confd process and rm service. Need to further investigate if that would somehow make rconfc miss that the confd is already down. Probably some process state checking is missing.
But I think, let's complete this PR and investigate this issue in another ticket.

Reply:
This will be investigated further in motr  https://jts.seagate.com/browse/EOS-20589 and from Hare point of as well.